pkg/thanos/statefulset.go

```diff
@@ -0,0 +1,337 @@
+	}
+
+	if old != nil {
+		statefulset.Annotations = old.Annotations
+	}
+
+	storageSpec := tr.Spec.Storage
+	if storageSpec == nil {
+		statefulset.Spec.Template.Spec.Volumes = append(statefulset.Spec.Template.Spec.Volumes, v1.Volume{
+			Name: volumeName(tr.Name),
+			VolumeSource: v1.VolumeSource{
+				EmptyDir: &v1.EmptyDirVolumeSource{},
+			},
+		})
+	} else if storageSpec.EmptyDir != nil {
+		emptyDir := storageSpec.EmptyDir
+		statefulset.Spec.Template.Spec.Volumes = append(statefulset.Spec.Template.Spec.Volumes, v1.Volume{
+			Name: volumeName(tr.Name),
+			VolumeSource: v1.VolumeSource{
+				EmptyDir: emptyDir,
+			},
+		})
+	} else {
+		pvcTemplate := storageSpec.VolumeClaimTemplate
+		pvcTemplate.CreationTimestamp = metav1.Time{}
+		if pvcTemplate.Name == "" {
+			pvcTemplate.Name = volumeName(tr.Name)
+		}
+		if storageSpec.VolumeClaimTemplate.Spec.AccessModes == nil {
+			pvcTemplate.Spec.AccessModes = []v1.PersistentVolumeAccessMode{v1.ReadWriteOnce}
+		} else {
+			pvcTemplate.Spec.AccessModes = storageSpec.VolumeClaimTemplate.Spec.AccessModes
+		}
+		pvcTemplate.Spec.Resources = storageSpec.VolumeClaimTemplate.Spec.Resources
+		pvcTemplate.Spec.Selector = storageSpec.VolumeClaimTemplate.Spec.Selector
+		statefulset.Spec.VolumeClaimTemplates = append(statefulset.Spec.VolumeClaimTemplates, pvcTemplate)
+	}
+
+	for _, volume := range tr.Spec.Volumes {
+		statefulset.Spec.Template.Spec.Volumes = append(statefulset.Spec.Template.Spec.Volumes, volume)
+	}
+
+	return statefulset, nil
+}
+
+func makeStatefulSetSpec(tr *monitoringv1.ThanosRuler, config Config, ruleConfigMapNames []string) (*appsv1.StatefulSetSpec, error) {
+	// Before editing 'tr' create deep copy, to prevent side effects. For more
+	// details see https://github.com/coreos/prometheus-operator/issues/1659
+	tr = tr.DeepCopy()
+
+	if len(tr.Spec.QueryEndpoints) < 1 {
+		return nil, errors.New(tr.GetName() + ": thanos ruler requires at least one query endpoint")
+	}
+
+	trCLIArgs := []string{
+		"rule",
+		fmt.Sprintf("--eval-interval=%s", "15s"),
+		fmt.Sprintf("--data-dir=%s", storageDir),
+	}
+	trEnvVars := []v1.EnvVar{}
+
+	if tr.Spec.ListenLocal {
+		trCLIArgs = append(trCLIArgs, "--http-address=localhost:10902")
+	}
+	if tr.Spec.LogLevel != "" && tr.Spec.LogLevel != "info" {
+		trCLIArgs = append(trCLIArgs, fmt.Sprintf("--log.level=%s", tr.Spec.LogLevel))
+	}
+	if tr.Spec.LogFormat != "" {
+		trCLIArgs = append(trCLIArgs, fmt.Sprintf("--log.format=%s", tr.Spec.LogFormat))
+	}
+	for _, endpoint := range tr.Spec.QueryEndpoints {
+		trCLIArgs = append(trCLIArgs, fmt.Sprintf("--query=%s", endpoint))
+	}
+	for _, ruleConfigMapName := range ruleConfigMapNames {
+		rulePath := rulesDir + "/" + ruleConfigMapName + "/*.yaml"
+		trCLIArgs = append(trCLIArgs, fmt.Sprintf("--rule-file=%s", rulePath))
+	}
+
+	if tr.Spec.AlertManagersConfig != nil {
+		trCLIArgs = append(trCLIArgs, "--objstore.config=$(ALERTMANAGERS_CONFIG)")
```

Comment:
`--alertmanagers.config` instead of `--objstore.config`

Reply:
Done